hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java

```diff
@@ -399,6 +400,44 @@ public RegisteredCommand register(
         LOG.error("Cannot find datanode {} from nodeStateManager",
             datanodeDetails.toString());
       }
+    } else {
+      // Update datanode if it is registered but the ip or hostname changes
+      try {
+        final DatanodeInfo datanodeInfo =
+                nodeStateManager.getNode(datanodeDetails);
+        if (!datanodeInfo.getIpAddress().equals(datanodeDetails.getIpAddress())
+                || !datanodeInfo.getHostName()
+                .equals(datanodeDetails.getHostName())) {
+          LOG.info("Updating data node {} from {} to {}",
+                  datanodeDetails.getUuidString(),
+                  datanodeInfo,
+                  datanodeDetails);
+          if (clusterMap.contains(datanodeInfo)) {
```

Comment:
It might be better to implement `clusterMap.update(datanodeDetails)`. This would keep the locking and concurrency issues in check.

Reply:
updated